process_pr.py

```diff
@@ -1254,38 +1348,120 @@ def process_pr(repo_config, gh, repo, issue, dryRun, cmsbuild_user=None, force=F
-                        (repository in auto_test_repo) or ("*" in auto_test_repo)
-                    ):
-                        test_comment = comment
-                    if sign == "orp":
-                        mustClose = False
-            elif ctype == "-1":
-                for sign in selected_cats:
-                    signatures[sign] = "rejected"
-                    if sign == "orp":
-                        mustClose = False
+    if (
+        (not warned_too_many_commits)
+        and not ok_too_many_commits
+        and pr.commits >= MAX_INITIAL_COMMITS_IN_PR
+    ):
+        if not dryRun:
+            issue.create_comment(
+                f"This PR contains too many commits ({pr.commits} > {MAX_INITIAL_COMMITS_IN_PR}). "
+                "Make sure you chose the right target branch.\n"
+                f"{', '.join([gh_user_char + name for name in CMSSW_ISSUES_TRACKERS])}, "
+                "you can override this check with `+commit-count`."
+            )
+        return
 
-    # end of parsing comments section
+    api_rate_limits(gh)
+
+    # Get the commit cache from `already_seen` commit or technical commit
+    print("Checking for commit cache")
+    cache_comment = None
+
+    if technical_comment:
+        cache_comment = technical_comment
+    else:
+        if already_seen:
+            cache_comment = already_seen
+
+    if cache_comment:
+        seen_commits_match = REGEX_COMMITS_CACHE.search(cache_comment.body)
+        if seen_commits_match:
+            print("Loading commit cache")
+            commit_cache = loads(seen_commits_match[1])
+
+    if pr.commits < MAX_INITIAL_COMMITS_IN_PR or ok_too_many_commits:
+        for commit in all_commits:
+            if commit.sha not in commit_cache:
+                commit_cache[commit.sha] = {
+                    "time": int(commit.commit.committer.date.timestamp()),
+                    "files": sorted(get_changed_files_in_commit(repo, commit)),
+                }
+
+            cache_entry = commit_cache[commit.sha]
+            events[datetime.fromtimestamp(cache_entry["time"])] = {
+                "type": "commit",
+                "value": cache_entry["files"],
+            }
+
+    print("Saving commit cache")
+    old_body = cache_comment.body if cache_comment else CMSBOT_TECHNICAL_MSG
+    new_body = (
+        REGEX_COMMITS_CACHE.sub("", old_body)
+        + "<!-- commits cache: "
+        + dumps_compact(commit_cache)
+        + " -->"
+    )
+    if not dryRun:
+        if cache_comment:
+            if old_body != new_body:
+                cache_comment.edit(new_body)
+        else:
+            issue.create_comment(new_body)
+    else:
+        if cache_comment:
+            print("DRY RUN: Updating existing comment with text")
+            print(new_body)
+        else:
+            print("DRY RUN: Creating technical comment with text")
+            print(new_body)
+
+    events = dict(sorted(events.items()))
```

Comment:
we should not depend on the `dict` ordering here. As mentioned [here](https://docs.python.org/3.6/whatsnew/3.6.html#new-dict-implementation)
```
The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon (this may change in the future, but it is desired to have this new dict implementation in the language for a few releases before changing the language spec to mandate order-preserving semantics for all current and future Python implementations; this also helps preserve backwards-compatibility with older versions of the language where random iteration order is still in effect, e.g. Python 3.5).
```

I would suggest to remove this and use something like `sorted(list(events.keys()))` directly in the `for` loop at line 1427

Reply:
On the other hand https://docs.python.org/3/library/stdtypes.html#dict says
> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was an implementation detail of CPython from 3.6.

(I checked because I think we are already relying on the ordering behavior in several places in CMSSW)

(but maybe the sorted list of tuples is better anyway)